### PR TITLE
QByteArray

### PR DESCRIPTION
**Pull request by shelltitan**
#214

### CODE REVIEW

**Comment by kremius** [comment]
Have you tested your changes with multiple clients and unsync generation?
It is needed to launch the master client with `-unsync_generation` flag & connect another client to server too. After that type `/generate_unsync` into the chat window and artificial unsync event will be generated -> the server should generate two dumps. They should contain at the end the list of last messages.

**Comment by shelltitan** [comment]
Yep it works fine, I did /generate_unsync and I got 2 dumps. I could connect 2 client at same time and play without unsync.

**Comment by kremius** [comment]
Do these two dumps contain a list of messages in the end? It is needed to unpack them with QzipUnpacker in order to look into them.

**Comment by shelltitan** [comment]
[1.txt](https://github.com/griefly/griefly/files/587768/1.txt)
[2.txt](https://github.com/griefly/griefly/files/587769/2.txt)

File: sources/core/Game.cpp [hunk 1]
```diff
@@ -707,15 +707,15 @@ void Game::generateUnsync()
...
-    stream << std::endl;
+    data.append('\n');
     for (int i = (log_pos_ + 1) % messages_log_.size();
              i != log_pos_;
              i = (i + 1) % messages_log_.size())
     {
-        stream << messages_log_[i].type << " ";
-        stream << messages_log_[i].json.toStdString() << std::endl;
+        data.append(messages_log_[i].type + " ");
+        data.append(messages_log_[i].json + '\n');
```
Comment (kremius) [inline sources/core/Game.cpp]: The dumps show that there is an error.
`" "` is actually a pointer, and `messages_log_[i].type` is a number, and you can do the pointer arithmetic in C++. So when you do that you get pointer in some memory area, and some random data is written into `data`.
It should be `QByteArray::number(messages_log_[i].type) + " "` instead.

If you fix that then remake the dumps test.

Comment (shelltitan) [inline sources/core/Game.cpp]: Isn't " " a constant string? Sorry I cannot imagine how " " can be a pointer, so I'm curious. I knew about pointer arithmetic just didn't know " " is a pointer in this case.

Comment (kremius) [inline sources/core/Game.cpp]: The actual type of `" "` is `const char[2]` (constant array of chars, the size is two).
But plain arrays in C++ is actually just constant pointers (there are some differences, but not so much), so string literals in C++ (`"something something"`) in a wide amount of cases behave like constant pointers (like a `const char*` pointer).

Comment (shelltitan) [inline sources/core/Game.cpp]: Thank you for your detailed explanation!

**Comment by shelltitan** [comment]
[4.txt](https://github.com/griefly/griefly/files/587810/4.txt)
[3.txt](https://github.com/griefly/griefly/files/587811/3.txt)

**Comment by kremius** [comment]
Yeah. now everything seems fine.

**Pull request merged**